Public/Src/Sandbox/Windows/DetoursServices/DetouredFunctions.cpp

```diff
@@ -2085,10 +2085,11 @@ HANDLE WINAPI Detoured_CreateFileW(
     // read request which may or may not have been approved (due to special exceptions for directories and non-existent files).
     // It is safe to go ahead and perform the real CreateFile() call, and then to reason about the results after the fact.
 
-    // Note that we can add FILE_SHARE_DELETE to dwShareMode. I.e., in order to leverage NTFS hardlinks to avoid copying cache
-    // content, we need to be able to delete one of many links to a file. Unfortunately, share-mode is aggregated only per file
+    // Note that we need to add FILE_SHARE_DELETE to dwShareMode to leverage NTFS hardlinks to avoid copying cache

```

Comment:
>    // N [](start = 0, length = 8)

There's a slight chance that we may still run into problems here. The QTest sandbox is created with CopyFile pips, which use hardlinks by default. So conceivably a test could have a handle open to a file in its sandbox while BuildXL tries to delete another hardlink to the same file that's used elsewhere in the build. In all likelihood I don't think we'll run into that, but it'll be a pain to debug if we do.

I guess we could only allow this when files are copied rather than hardlinked. That could be config based or you could check the link count of files. Doing the latter may have perf implications.

Again, I think we'll probably be fine with this. Just wanted to bring this up in case you think it's worth addressing now.

Reply:
Thanks for the scenario. I think we should have a unit test for this.